pkg/verification/service/verification_service.go

```diff
@@ -256,28 +303,72 @@ func (s *ServiceImpl) VerifyCode(ctx *gin.Context, userID string, code string) (
 		if code != signup.Annotations[toolchainv1alpha1.UserSignupVerificationCodeAnnotationKey] {
 			// The code doesn't match
 			attemptsMade++
-			signup.Annotations[toolchainv1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
+			annotationValues[toolchainv1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
 			verificationErr = errors.NewForbiddenError("invalid code", "the provided code is invalid")
 		}
 	}
 
 	if verificationErr == nil {
 		// If the code matches then set VerificationRequired to false, reset other verification annotations
-		states.SetVerificationRequired(signup, false)
-		delete(signup.Annotations, toolchainv1alpha1.UserSignupVerificationCodeAnnotationKey)
-		delete(signup.Annotations, toolchainv1alpha1.UserVerificationAttemptsAnnotationKey)
-		delete(signup.Annotations, toolchainv1alpha1.UserSignupVerificationCounterAnnotationKey)
-		delete(signup.Annotations, toolchainv1alpha1.UserSignupVerificationInitTimestampAnnotationKey)
-		delete(signup.Annotations, toolchainv1alpha1.UserVerificationExpiryAnnotationKey)
+		unsetVerificationRequired = true
+		annotationsToDelete = append(annotationsToDelete, toolchainv1alpha1.UserSignupVerificationCodeAnnotationKey)
+		annotationsToDelete = append(annotationsToDelete, toolchainv1alpha1.UserVerificationAttemptsAnnotationKey)
+		annotationsToDelete = append(annotationsToDelete, toolchainv1alpha1.UserSignupVerificationCounterAnnotationKey)
+		annotationsToDelete = append(annotationsToDelete, toolchainv1alpha1.UserSignupVerificationInitTimestampAnnotationKey)
+		annotationsToDelete = append(annotationsToDelete, toolchainv1alpha1.UserVerificationExpiryAnnotationKey)
 	} else {
 		log.Error(ctx, verificationErr, "error validating verification code")
 	}
 
-	// Update changes made to the UserSignup
-	_, updateErr := s.Services().SignupService().UpdateUserSignup(signup)
-	if updateErr != nil {
-		log.Error(ctx, updateErr, "error updating UserSignup")
-		verificationErr = errors.NewInternalError(updateErr, "error updating UserSignup")
+	doUpdate := func() error {
+		signup, err := s.Services().SignupService().GetUserSignup(userID)
+		if err != nil {
+			return err
+		}
+
+		if unsetVerificationRequired {
+			states.SetVerificationRequired(signup, false)
+		}
+
+		for k, v := range annotationValues {
+			signup.Annotations[k] = v
+		}
+
+		for _, annotationName := range annotationsToDelete {
+			delete(signup.Annotations, annotationName)
+		}
+
+		_, err = s.Services().SignupService().UpdateUserSignup(signup)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// Attempt to update the UserSignup, retrying a number of times if the update fails
+	attempts := 0
+	for {
```

Comment:
Looks good! But maybe it would look a bit cleaner if we introduce some function like:
```
func PollUpdateSignup(updater func() error) error {
   ...
}
```
And reuse it in both cases?

Reply:
No problem, I've added something like this.